Updated native module setup docs for 61, 62, 63, and master (#268)
microsoft/react-native-windows-samples: docs/native-modules-setup.md

Make sure you have installed all of the [development dependencies](rnw-dependencies.md).

> If you're planning on writing in C++, you *must* install the [C++/WinRT Visual Studio Extension](https://marketplace.visualstudio.com/items?itemName=CppWinRTTeam.cppwinrt101804264).

Reviewer: shouldn't we keep this on? or at least make it a recommendation?

Reviewer: actually it might make sense for rnw-dependencices to install it

Author: AFAIK this only adds the C++/WinRT project templates to VS. Since the CLI is creating the project for them, they don't really need it.

Follow the official React Native instructions at https://reactnative.dev/docs/native-modules-setup.

```bat
npx create-react-native-module --module-name "MyLibrary" MyLibrary
cd MyLibrary
yarn install
```

Now you'll have a new native module project under `MyLibrary`. Be sure to look at the command output for further steps you'll want to do before publishing the project.

At this point, follow the steps below to add Windows support to the newly created library.

### Updating your package.json

Unfortunately the default library template targets older versions of `react` and `react-native` that Windows supports, so you'll need to upgrade to newer versions of `react` and `react-native` in order to add supprot for `react-native-windows`.

> For more information on how NPM dependencies work, see [Specifying dependencies and devDependencies in a package.json file](https://docs.npmjs.com/specifying-dependencies-and-devdependencies-in-a-package-json-file).

> You can use the `npm info` command to find the correct versions to use. Let's assume you plan on building against the latest `canary` version of `react-native-windows`.

Reviewer 2: Might be better to point to `latest` instead of canary for examples, since Canary has weird version numbers, etc.

Author: Each version of the doc points to the version that doc is for. The files under `docs` are for the "next" release, after `latest`, which right now is `canary`. Then the next preview starts I'll change it to `preview`.

Author: BTW, is there any way we can "unpublish" `preview` from npm when there's no active preview? Or re-point `preview` to latest once we officially release? It's a little weird when `preview` predates `latest`.

Reviewer 2: We can remove tags if we want to. `react-native` uses `next` instead of `preview`, and kind of does the same thing where `next precedes` `latest` in certain cases (I think there's is honestly more confusing).

Author: We can take this offline.

Use the following command to find the matching versions of `react`:

```bat
npm info react-native-windows@canary devDependencies.react
```

Take the result of that command (let's say it's `16.13.1`) and use it to upgrade the dev dependency:

```bat
yarn upgrade react@16.13.1 --dev
```

You'll need to repeat the steps for `react-native`, ie:

```bat
npm info react-native-windows@canary devDependencies.react-native
```

Again, take the result of that command (let's say it's `0.0.0-a36d9cd7e`) and use it to upgrade the dev dependency:

```bat
yarn upgrade react-native@0.0.0-a36d9cd7e --dev
```

Now you should be ready to add Windows support with the CLI.

> ## Adding Windows support to an existing library

Reviewer 2: Does `create-react-native-module` add example projects for iOS/Android that would need to be revved?

Author: `create-react-native-module` *can* create example projects with a flag.

I'm not sure what guidance to give about the main project's dependencies, let alone example app dependencies. The instructions here make sure that you have the right files on disk to build. And even if you publish to npm (with or without the changes to package.json), an RNW app will be able to consume your native module.

This is least intrusive way to implement modules without any assumptions or proscriptions about the module's maintenance.

Reviewer 2: My concern was more that someone could read this, then rev existing examples versions without necessarily keeping them up to date. I.e. if we create iOS/Android examples for a version, then run these commands, there is some extra work to do still that has tripped people up before.

Could we add a sentence pointing folks towards upgrade-helper if there are a existing sample applications?

> Similar to adding Windows support to an app project, you can use the CLI to add Windows support to a native module project, but you'll need to specify `--projectType lib`:

Reviewer 2: I'm not sure if there's a good solution, but I remember confusion before when someone tried running commands in our docs at the root of a monorepo. Not sure if adding something about "run this in the directory with your module package" would be useful or not.

Author: I can make the language clearer that the location to run the command is usually the repo root but not always.

```bat
npx react-native-windows-init --version canary --projectType lib --overwrite
```

This defaults to a C++/WinRT project. If you want to create a C# based native module project, use:

```bat

> npx react-native-windows-init --version canary --projectType lib --language cs --overwrite

Reviewer 2: What does `react-native-windows-init` lib mode do for dependencies/peer dependencies? Feels like we might need to do something manually here. I.e. we don't want any deps on react-native-windows, but do want a permissive peer dependency.

Author: Th logic is here: https://github.com/microsoft/react-native-windows/blob/master/packages/react-native-windows-init/src/Cli.ts#L331

From what I can see, only apps depend directly on react, react-native, and/or react-native-windows. Libraries may have zero dependencies, and at minimum just have react-native as a dev dependency.

So if I see dependencies with react-native (probably an app), we use the previous logic of inserting rnw into dependencies and running yarn/npm install.

If I see 'react-native' as a dev dependency (probably a lib), then add rnw as a dev dependency with yarn/npm

Reviewer 2: Makes sense. I think we need to add a step asking to add a peer dependency then, since the CLI just adds a dev dependency.

Author: I think there needs to some "global" statement that if you're adding windows implementation to a community module, you'll need to work with the module owners when making any `package.json` changes.

Follow the official React Native instructions at https://reactnative.dev/docs/native-modules-setup.

```bat
npx create-react-native-module --module-name "MyLibrary" MyLibrary
cd MyLibrary
yarn install
```

Now you'll have a new native module project under `MyLibrary`. Be sure to look at the command output for further steps you'll want to do before publishing the project.

At this point, follow the steps below to add Windows support to the newly created library.

### Updating your package.json

> Unfortunately the default library template targets older versions of `react` and `react-native` that Windows supports, so you'll need to upgrade to newer versions of `react` and `react-native` in order to add supprot for `react-native-windows`.

Reviewer: ```suggestion
Unfortunately the default library template targets older versions of `react` and `react-native` that Windows supports, so you'll need to upgrade to newer versions of `react` and `react-native` in order to add support for `react-native-windows`.
```

> Properly defining your NPM dependencies is an essential part of creating and maintaining a React Native library, especially one that supports multiple platforms. The instructions here represent the minimum steps required to start targeting `react-native-windows`. If you're adding Windows support to a library you don't own, you'll need to work with the library owners to make sure any changes made to `package.json` are appropriate.
> 
> For more information on how NPM dependencies work, see [Specifying dependencies and devDependencies in a package.json file](https://docs.npmjs.com/specifying-dependencies-and-devdependencies-in-a-package-json-file).

You can use the `npm info` command to find the correct versions to use. Let's assume you plan on building against the latest `canary` version of `react-native-windows`.

Use the following command to find the matching versions of `react`:

```bat
npm info react-native-windows@canary devDependencies.react
```

Take the result of that command (let's say it's `16.13.1`) and use it to upgrade the dev dependency:

```bat
yarn upgrade react@16.13.1 --dev
```

> You'll need to repeat the steps for `react-native`, ie:

Reviewer: ```suggestion
You'll need to repeat the steps for `react-native`, i.e.:
```

```bat
npm info react-native-windows@canary devDependencies.react-native
```

Again, take the result of that command (let's say it's `0.0.0-a36d9cd7e`) and use it to upgrade the dev dependency:

```bat
yarn upgrade react-native@0.0.0-a36d9cd7e --dev
```

Now you should be ready to add Windows support with the CLI.

## Adding Windows support to an existing library

> Similar to adding Windows support to an app project, you can use the CLI to add Windows support to a native module project, but you'll need to specify `--projectType lib`:

Reviewer: ```suggestion
Similarly to adding Windows support to an app project, you can use the CLI to add Windows support to a native module project, by specifying `--projectType lib`:
```

### Making your module ready for consumption in an app

If you've followed the steps above, your module should be ready for consumption thanks to [Autolinking](native-modules-autolinking.md).

However, there are some things you may need to check:

1. If you are writing a C++/WinRT module and have added any NuGet package dependencies, you'll see references to those packages in your vcxproj file as relative references e.g. `..\packages\...`. We need these to use the solution directory instead, so replace all mentions of `..\packages\` with `$(SolutionDir)\`.

Example, change this:

```xml
  <Import Project="..\packages\NuGetPackage.1.0.0.0\build\native\NuGetPackage.props" Condition="Exists('..\packages\NuGetPackage.1.0.0.0\build\native\NuGetPackage.props')" />
```

> to this:

Reviewer: consider using the diff format (```diff)
